File: src/fs/IUBH.TOR/IUBH.TOR/CoursePageHtmlParser.fs
module IUBH.TOR.CoursePageHtmlParser

open FSharp.Data
open System

/// Returns the "inner text" of a table cell (<td />).
let private td (index : int) (row : HtmlNode) =
    let node = row.Descendants [ "td" ] |> Seq.item index
    let text = (HtmlNode.innerText node)
    if text = null then text else text.Trim()

/// Creates a Course out of a table row
let private mapRawCourse (moduleName, row) =
    { Id = (row |> td 1).GetHashCode()
      Title = row |> td 1
      Module = moduleName
      Status = row |> td 2
      Grade = row |> td 3
      Rating = row |> td 4
      Credits = row |> td 5
      DateOfExamination = row |> td 6
      Attempts = row |> td 8 }

/// Deterines if a given row is a module row
let private isModuleRow (row : HtmlNode) =
    /// Module rows don't contain links
    row.Descendants [ "a" ] |> Seq.isEmpty
    
/// Determines if a course has an examination date
let private hasDateOfExamination rawCourse =
    not (String.IsNullOrWhiteSpace rawCourse.DateOfExamination)

/// Tries to parse the HTML document, provided as a
/// string argument. If the doc could be parsed, a
/// list of courses is being returned. If not, an
/// error message is being given back instead.
let tryParse htmlDocument =
    try
        /// We are using the F# HTML Type Provider to parse the doc
        let htmlDocument = HtmlDocument.Parse htmlDocument
        
        /// The doc is structured in different tables. We collect
        /// Them to get all the courses and modules in a flat list.
        let rows =
            htmlDocument.Descendants "tbody"
            |> Seq.rev
            |> Seq.skip 1 // The last one isn't a semester table
            |> Seq.collect (fun table -> table.Descendants [ "tr" ])
            |> Seq.toList

        /// The result must contain
        /// - All courses
        /// - But not those without examination date
        /// - The module name of each course
        /// - No modules
        let result =
            rows
            |> Seq.mapi (fun index row ->
                if not (row |> isModuleRow) then
                   let moduleName =
                       rows.[..index]
                       |> Seq.findBack isModuleRow
                       |> td 1
                   Some(moduleName, row)
                else None)
            |> Seq.filter (fun x -> x.IsSome)
            |> Seq.map (fun x -> x.Value |> mapRawCourse)
            |> Seq.filter (hasDateOfExamination)
            |> Seq.toList

        Ok result
    with e -> Error e.Message 

File: src/cs/IUBH.TOR/IUBH.TOR/Modules/Courses/Services/CoursePageHtmlParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using HtmlAgilityPack;
using IUBH.TOR.Domain;
using IUBH.TOR.Modules.Courses.Domain;

namespace IUBH.TOR.Modules.Courses.Services
{
    internal class CoursePageHtmlParser : ICoursePageHtmlParser
    {
        /// <summary>
        /// Converts an HTML page to a set of (raw) course information.
        /// This is tailored towards the current structure of the HTML
        /// document. So it is likely to break as soon as that structure
        /// changes. Use the Unit Tests provided to adjust the mechanics
        /// of this method in that case.
        /// </summary>
        public Result<RawCourse[]> TryParseCoursePage(string coursePageHtml)
        {
            try
            {
                var doc = new HtmlDocument();
                doc.LoadHtml(coursePageHtml);

                // Collects all tables on the page.
                var tables = doc.DocumentNode
                    .SelectNodes("//tbody").Reverse()
                    .Skip(1); // The last one isn't a semester table

                var tableRows = tables.SelectMany(t => t.ChildNodes);

                var courses = new List<RawCourse>();

                string currentModuleName = null;

                foreach (HtmlNode tableRow in tableRows)
                {
                    var columns = tableRow.ChildNodes;

                    if (columns.Count == 0)
                    {
                        continue;
                    }

                    // Skipping modules – those modules don't contain links
                    bool isModule = !tableRow.Descendants("a").Any();

                    // But if we have a module here, we remember its name so
                    // we can pass it along all subsequent courses until the
                    // next module appears
                    if (isModule)
                    {
                        currentModuleName = cleanString(columns[1].InnerText);
                        continue;
                    }

                    string dateOfExamination = columns[6].InnerText;

                    // If there's no date of examination this course information
                    // is not of interest to us. So we skip it.
                    if (string.IsNullOrWhiteSpace(dateOfExamination))
                    {
                        continue;
                    }

                    // To avoid side effects in the representation later on it's
                    // important to strip all characters we don't need from the texts.
                    string cleanString(string input)
                        => Regex.Replace(input ?? string.Empty, @"(\n|\r|\t|\s)+", " ").Trim();

                    string title = cleanString(columns[1].InnerText);
                    string id = title.GetHashCode().ToString();
                    
                    courses.Add(
                        new RawCourse
                        {
                            Id = id,
                            Title = title,
                            Module = currentModuleName,
                            Status = cleanString(columns[2].InnerText),
                            Grade = cleanString(columns[3].InnerText),
                            Rating = cleanString(columns[4].InnerText),
                            Credits = cleanString(columns[5].InnerText),
                            DateOfExamination = cleanString(dateOfExamination),
                            Attempts = cleanString(columns[8].InnerText)
                        }
                    );
                }

                return Result.WithSuccess(courses.ToArray());
            }
            catch (Exception e)
            {
                return Result.WithException<RawCourse[]>(e);
            }
        }
    }
}
